### PR TITLE
Expand assert!(expr, args..) to include $crate for hygiene on 2021.

### PR DESCRIPTION
**Pull request by m-ou-se**
This makes `assert!(expr, args..)` properly hygienic in Rust 2021.

This is part of rust-lang/rfcs#3007, see #80162.

Before edition 2021, this was a breaking change, as `std::panic` and `core::panic` are different. In edition 2021 they will be identical, making it possible to apply proper hygiene here.

### CODE REVIEW

**Comment by m-ou-se** [comment]
(cc @petrochenkov, as this touches a (builtin) macro)

File: compiler/rustc_builtin_macros/src/assert.rs [hunk 1]
```diff
@@ -23,16 +23,34 @@ pub fn expand_assert<'cx>(
         }
     };
 
+    let is_2021 = sp.rust_2021();
```
Comment (petrochenkov) [inline compiler/rustc_builtin_macros/src/assert.rs:26]: Edition checks are generally expensive, so it's better to not perform them unless the result is actually used.
So it's better to move this to inside the `custom_message` condition (note that edition checking should use the `sp` before `with_call_site_ctxt`).

**Comment by petrochenkov** [comment]
r=me after addressing the comment.

**Comment by m-ou-se** [comment]
The apple-alt CI runner is being slow again. This will time out.

@bors retry

**Pull request merged**